Code/GraphMol/Descriptors/WHIM.cpp

```diff
@@ -0,0 +1,625 @@
+}
+
+double* getWhimDesc(JacobiSVD<MatrixXd> svd, MatrixXd Xmean, int numAtoms, double th,bool printscore) {
+
+    double *w = new double[18];
+    // prepare data for Whim parameter computation
+    
+    double *SingVal = retreiveVect(svd.singularValues());
+    MatrixXd Scores= Xmean*svd.matrixV(); //  V is similar
+
+    // compute parameters
+    w[0]=SingVal[0];
+    w[1]=SingVal[1];
+    w[2]=SingVal[2];
+    w[3]=SingVal[0]+SingVal[1]+SingVal[2];  // T
+    w[4]=SingVal[0]*SingVal[1]+SingVal[0]*SingVal[2]+SingVal[1]*SingVal[2]; // A
+    w[5]=w[3]+w[4]+SingVal[0]*SingVal[1]*SingVal[2]; // V
+    w[6]=SingVal[0]/(SingVal[0]+SingVal[1]+SingVal[2]); // P1
+    w[7]=SingVal[1]/(SingVal[0]+SingVal[1]+SingVal[2]); // p2
+    w[8]=SingVal[2]/(SingVal[0]+SingVal[1]+SingVal[2]); // P3
+
+    double res=0.0;
+    for (int i=0;i<3;i++) 
+    {
+      res+=std::abs(w[i]/w[3]-1.0/3.0);
+    }
+
+    w[9]=3.0/4.0*res; // K
+
+    // center original matrix
+
+    VectorXd v1= Scores.col(0);
+    VectorXd v2= Scores.col(1);
+    VectorXd v3= Scores.col(2);
+
+    if (printscore and th ==0.03) {
+          std::cout << v1.transpose() << "\n";
+          std::cout << v2.transpose() << "\n";
+          std::cout << v3.transpose() << "\n";
+    }
+
+    w[10] = numAtoms*pow(w[0],2)/ v1.array().pow(4).sum(); // E1
+    w[11] = numAtoms*pow(w[1],2)/ v2.array().pow(4).sum(); // E2
+    w[12] = numAtoms*pow(w[2],2)/ v3.array().pow(4).sum(); // E3
+    w[13] = w[10]+w[11]+w[12]; //D
+
+
+    double gamma[3]; // Gamma values
+    double nAT = (double) numAtoms;
+    // check if two atoms are symetric versus the new axis ie newx,newy,newz a
+    for (int i=0;i<3; i++) {
+      for (int j =0;j<numAtoms;j++) {
+        Scores(j,i)=roundn(Scores(j,i),2); // round the matrix! same as eigen tolerance !
+      }
+    }
+
+    //std::cout << Scores.transpose() <<"\n";
+
+    for (int i=0;i<3; i++) {
+      double ns=0.0;
+      for (int j=0;j< numAtoms;j++) {
+        for (int k=0;k< numAtoms;k++){
+
+          if (j==k) continue;
+
+          if (std::abs(Scores(j,i) + Scores(k,i))< th and (std::abs(Scores(j,i))>0 or std::abs(Scores(k,i))>0 )) {  // those that are close opposite & not close to the axis!
+            ns++; // check only once the symetric none null we need to add +2!
+            break;
+          }
+        }
+      }
+
+      // for all atoms close to the axis we need to add +1!
+      for (int j=0;j< numAtoms;j++) {
+        if (Scores(j,i)==0) {
+        ns++; // atom close to the the axis are symetric! 
+        } 
+      }
+
+      double na=0.0;
+      na = nAT-ns;
+      gamma[i] =0.0;
+      if (ns>0) {
+           gamma[i] = ((ns / nAT) * log(ns / nAT) / log(2.0) + (na / nAT) * log(1.0 / nAT) / log(2.0));
+           gamma[i] = 1.0 / (1.0 - gamma[i]);
+      }
+    }
+    w[14]=gamma[0]; // G1
+    w[15]=gamma[1]; // G2
+    w[16]=gamma[2]; // G3
+    w[17]=pow(gamma[0]*gamma[1]*gamma[2],1.0/3.0);
+
+    return w;
+
+}
+
+
```

Comment:
Duplicated code that can be refactored into one generic function `GetWHIMweighted` and specific ones that calculated the weight and call `GetWHIMweighted`.